src/command_executor.cc

```diff
@@ -87,22 +89,28 @@ int CommandExecutor::Execute(const std::string &line) {
   std::vector<CommandParseResult> command_parse_result_list =
       parser_.ParseUserInputLine(line);
   int save_fd = dup(STDOUT_FILENO);
-  bool is_pipe_redirect = command_parse_result_list.size() > 1;
   std::shared_ptr<CommandParseResult> built_In_Command_ptr = nullptr;
-  int pipe_fds[2];
-  if (pipe(pipe_fds) == -1) {
-    utils::PrintSystemError(std::cerr);
-    return ERROR_CODE_DEFAULT;
+
+  auto pipe_number = command_parse_result_list.size() - 1;
+  std::vector<int *> pipe_fds_list;
+  for (int i = 0; i < pipe_number; i++) {
+    int *pipe_fds = new int[2];
```

Comment:
Memory leak, you use `new` to allocation some memory in heap, but not `delete` it.

Reply:
I'm thinking if we can use `array` and `vector` and to let STL manage memory itself. For example:

```
  std::vector<std::array<int, 2>> pipe_fds_list;
  for (int i = 0; i < pipe_number; i++) {
    int pipe_fds[2];
    if (pipe(pipe_fds) == -1) {
      utils::PrintSystemError(std::cerr);
      return ERROR_CODE_DEFAULT;
    }
    pipe_fds_list.push_back({pipe_fds[0], pipe_fds[1]});
  }
```